Temporary entry points in ngen

vitek-karas opened the pull request:
Doing this in a separate branch to make the PR easier to manage. Eventually this should be merged with the rest of the NGEN Precode work (as it's based on it).

- Save precodes as temporary entry points in NGEN images
- Only uses fixup/stub precodes as they can't be reused at runtime anyway
- Changes to zap structure tracking to allow for internal pointers to saved structures
- Save the pointer (relative) to the temporary entry points in front of the method desc chunk (just like runtime layout)
- Revert changes to method PreStub logic to basically behave just like for runtime temporary entry points with only very small differences.
- Remove code used to save precodes which are not used anymore in NGEN images

vitek-karas committed:
Add the abitlity to store temporary entry points in NGEN on MD chunk.

Basically saves teh MD chunk with one additional pointer in front of it. That's where we will store the temporary entry points pointer (just like the runtime allocated MD chunks do).
In order to do this cleanly we need to be able to map the MD chunk pointer to an ZapInnerPtr node (and not the structure node which backs up the memory since that is starting eith the temporary entry point pointer).
This introduces the ZapInnerPtr nodes to lot of the DataImage Fixup and BindPointer infra. So adapt that infra to be able to handle inner ptr nodes correctly.

vitek-karas committed:
Save precodes as temporary entry points

- Change precode saving to use the exact same layout as temporary entry points
- This also means that we save only one type of precode (either fixup or stub depending on defines).
- Store the saved precodes as the value for the temporary entry points slot in the MD chunk

No runtime changes yet, other than the fact that precodes for methods are now comming from a different place. Saved methods are still treated as not having temporary entry points.

vitek-karas committed:
Save temporary entry points for all methods in a chunk

Change chunk method ordering to consider the need for precodes - with this we try to put methods which need precodes together and split chunks on that. So we should have chunks where all (or most) methods need precodes and chunk where none do.
Save precodes only for chunks where at least one method needs them.

Register all precodes as surrogates of the "new" MD, so that we can run fixup on them (easy lookup). Leave the current registration of precodes as surrogates of the "old" MD only for methods which need precodes. This is used to figure out if the method should use precode in its MT slot.

vitek-karas committed:
Clean up precode marking during MD save

vitek-karas committed:
Remove unnecessary special casing for zapped methods

Remove the special code to deal with zapped precodes in GetOrCreatePrecode and SetStableEntryPoint. In both cases we can now rely on the GetTemporaryEntryPoint to work correctly. The only exceptrion is GetOrCreatePrecode which still has to enfore new prcode creation over zapped precode, regardless of the type of the zapped precode.
Also removed the special code in DoBackpatch as it can also rely on GetTemporaryEntryPoint.

vitek-karas committed:
Cleanup - remove unused code for saving precodes

jkotas commented on src/vm/method.cpp:
```diff
@@ -3684,16 +3682,28 @@ void MethodDesc::SaveChunk::SaveOneChunk(COUNT_T start, COUNT_T count, ULONG siz
+    // This helps with not spreading the knowledge of the storage of the temporary entry points slot
+    // throughout the codebase (since several pieces need to be able to map into the chunks
+    // during fixups).
+    ZapNode * pNode = m_pImage->GetInnerPtr(pStructureNode, sizeof(MethodDescChunk::TemporaryEntryPointsSlot));
```
Can `m_pImage->BindPointer(pNewChunk, pNode, sizeof(MethodDescChunk::TemporaryEntryPointsSlot));` do the trick instead of this inner pointer gymnastic?

vitek-karas replied:
I actually started with that.. 
Unfortunately there are several places where we ignore the offset part of the pointer->node lookup. Specifically in FixupPrecode this would introduce a bit more mess into the code. So I added the innerptr node - which worked well, but then broken in other places - but overall it didn't feel that bad.
It's one or the other - more complexity in the infra, for the benefit of keeping "users" cleaner. And also unaware of the actual layout of the method desc chunk and more importantly it's ZapNode storage details.
Or - less complex infra, but more knowledge of zap node storage layout outside of the MD chunk.
I personally prefer the approach with the inner ptr, but if you think the other way would be better I can definitely switch it.

jkotas replied:
I think my preference would be to do it the other way. I think it is better to keep the magic in the places that need them, not push them into the infra piece.

jkotas commented on src/vm/precode.cpp:
```diff
@@ -836,75 +796,103 @@ static void SaveFixupPrecodeChunk(DataImage * image, MethodDesc ** rgMD, COUNT_T
-        pendingCount++;
-    }
+PVOID Precode::SaveChunk::Save(DataImage * image)
+{
```
I think you will hit a problem with remoting on desktop here.

RemotingPrecode takes early out in the case the method is intercepted for remoting. The method itself won't even get JITed or restored from NGen image. I do not think that it will be the case with these changes because of there won't be remoting precode to do the interception. We will hit the prestub that will try to JIT/restore the method. JITing/restoring of the method is a problem because of it will try to load additional dependencies that may fail. There are real-world scenarios that will break because of this.

vitek-karas replied:
Just so that I understand this - you're saying that for remoting we need to keep saving RemotingPrecodes even though they can't be patched (and thus if we did get to PreStub for them, we would have to create a new RemotingPrecode for the method anyway)... right?

vitek-karas replied:
Just curious how this works for runtime loaded stuff... since we can't really have remoting precodes in temporary entry points (even at runtime temporary entry points are fixup precodes (or stub or compact))...
Does it mean that for these methods we create a stable entry point with the RemotingPrecode?

jkotas replied:
>  you're saying that for remoting we need to keep saving RemotingPrecodes even though they can't be patched

Right. Or do surgery on prestub to just generate remoting precode for the remoting interception, without JITing the method.

> since we can't really have remoting precodes in temporary entry points

We do not hand out the temporary entrypoints for these. I think that the following piece in TryGetMultiCallableAddrOfCode takes care of it as side-effect:

```
    // Force the creation of the precode if we would eventually got one anyway
    if (MayHavePrecode())
        return GetOrCreatePrecode()->GetEntryPoint();
```

vitek-karas replied:
I implemented the ability to save a specific precode (remoting for now) to a given method on top of the temporary entry point (so the method will have both, with the specific one used in MT slots).

jkotas replied:
> so the method will have both, with the specific one used in MT slots

How is the slot patching going to work with this? The slot is going to point to the remoting precode. The patching expects the slot to point to the temporary entrypoint. So it won't ever get patched, I think.

jkotas replied:
Thinking about this some more - maybe the surgery on the prestub would be easier way to deal with this. It should be pretty minor:  You really just need to call `GetOrCreatePrecode` and return if you get called with transparent proxy methodtable.

vitek-karas replied:
Reverted the change for specific precodes and instead added a rather simple branch into DoPrestub.

vitek-karas committed:
PR feedback: Switch from using inner ptr to save MethodDescChunk to an approach based on offsets

Introduced some constant so that code outside of the MD chunk saving logic can use those instead of relying on knowledge of the actual layout of the nodes.
Removed the DataImage changes to support inner ptr pointers with the Fixup... methods as those are not needed anymore.

vitek-karas commented:
@jkotas could you please take a look at the last change around remoting precodes? There's no way to test this in CoreCLR so that's not the point. I'm interested if the approach seems reasonable to you.

vitek-karas committed:
Methods which need remoting precode will get one in prestub without running the rest of prestub.

jkotas commented on src/vm/prestub.cpp:
```diff
@@ -1714,6 +1693,22 @@ PCODE MethodDesc::DoPrestub(MethodTable *pDispatchingMT)
+    // Zapped methods which need remoting precode are saved with normal temporary entry point
+    // (so fixup or stub precode). So the first time we run into them here we need to create
+    // the remoting precode for them and patch everywhere with it.
+    if (IsZapped() && GetPrecodeType() == PRECODE_REMOTING)
```
This means that we will always take this path - we will never take the path that does the JITing.

I think the condition should be something like:

`if (!HasStableEntryPoint() && GetPrecodeType() == PRECODE_REMOTING)`

vitek-karas replied:
Right - my brain obviously didn't work well yesterday. Thanks for spotting this!

jkotas reviewed (approved):
LGTM modulo nit

vitek-karas committed:
Fix remoting precode creation in Prestub

Only force create remoting precode if the method doesn't have a stable entry point yet. This should really only happen for zapped methods, but it would work on all.
On subsequent calls to prestub for the same method, it would already have a stable entry point and thus the normal perstub logic would execute.
